[Flaking-Test] [sig-scheduling] test/integration/scheduler/serving.serving

### Which jobs are flaking?

sig-release-master-blocking
- integration-master

### Which tests are flaking?

`k8s.io/kubernetes/test/integration/scheduler/serving.serving`  `TestEndpointHandlers`
[Triage](https://storage.googleapis.com/k8s-triage/index.html?text=TestEndpointHandlers&test=k8s.io%2Fkubernetes%2Ftest%2Fintegration%2Fscheduler%2Fserving.serving&xjob=e2e-kops)
https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1887928819386421248

### Since when has it been flaking?

[1/31/2025, 7:16:04 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1885142155438067712)
[2/5/2025, 8:48:19 AM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1886977308703592448)
[2/4/2025, 12:56:17 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1886677329376186368)
[2/7/2025, 11:49:11 PM](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-integration-master/1887928819386421248)

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#integration-master

### Reason for failure (if possible)

```
{Failed  === RUN   TestEndpointHandlers//readyz_with_sched-handler-sync
        	apiVersion: kubescheduler.config.k8s.io/v1
        	clientConnection:
        	  acceptContentTypes: ""
        	  burst: 100
        	  contentType: application/vnd.kubernetes.protobuf
        	  kubeconfig: /tmp/kubeconfig174648122
        	  qps: 50
        	enableContentionProfiling: true
        	enableProfiling: true
        	kind: KubeSchedulerConfiguration
        	leaderElection:
        	  leaderElect: false
        	  leaseDuration: 15s
        	  renewDeadline: 10s
        	  resourceLock: leases
        	  resourceName: kube-scheduler
        	  resourceNamespace: kube-system
        	  retryPeriod: 2s
        	parallelism: 16
        	percentageOfNodesToScore: 0
        	podInitialBackoffSeconds: 1
        	podMaxBackoffSeconds: 10
        	profiles:
        	- pluginConfig:
        	  - args:
        	      apiVersion: kubescheduler.config.k8s.io/v1
        	      kind: DefaultPreemptionArgs
        	      minCandidateNodesAbsolute: 100
        	      minCandidateNodesPercentage: 10
        	    name: DefaultPreemption
        	  ------
        	    preEnqueue: {}
        	    preFilter: {}
        	    preScore: {}
        	    queueSort: {}
        	    reserve: {}
        	    score: {}
        	  schedulerName: default-scheduler
         >
    endpoints_test.go:179: expected http 200 at /readyz/sched-handler-sync of component, got: 500 "internal server error: waiting for handlers to sync\n"
--- FAIL: TestEndpointHandlers//readyz_with_sched-handler-sync (1.45s)

=== RUN   TestEndpointHandlers
    testserver.go:591: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:411: runtime-config=map[api/all:true]
    testserver.go:412: Starting kube-apiserver on port 34701...
    testserver.go:447: Waiting for /healthz to be ok...
[-]poststarthook/start-apiextensions-controllers failed: not finished
[-]poststarthook/crd-informer-synced failed: not finished
[-]poststarthook/start-service-ip-repair-controllers failed: not finished
-------
-------
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/priority-and-fairness-config-producer failed: not finished
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-02-07 19:16:00.422094455 +0000 UTC m=+2.981502501"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "system" and it exists`,
  }
--------
---------
  	Type:               "Dangling",
- 	Status:             "",
+ 	Status:             "False",
- 	LastTransitionTime: v1.Time{},
+ 	LastTransitionTime: v1.Time{Time: s"2025-02-07 19:16:00.555912642 +0000 UTC m=+3.115320528"},
- 	Reason:             "",
+ 	Reason:             "Found",
- 	Message:            "",
+ 	Message:            `This FlowSchema references the PriorityLevelConfiguration object named "catch-all" and it exists`,
  }
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
---
---
[-]poststarthook/scheduling/bootstrap-system-priority-classes failed: not finished
--- FAIL: TestEndpointHandlers (19.09s)
```

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig scheduling

--- a/test/integration/scheduler/serving/endpoints_test.go
+++ b/test/integration/scheduler/serving/endpoints_test.go
@@ -29,6 +29,7 @@ import (
 	"testing"
 
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/util/retry"
 	featuregatetesting "k8s.io/component-base/featuregate/testing"
 	"k8s.io/component-base/zpages/features"
 	"k8s.io/klog/v2/ktesting"
@@ -175,29 +176,46 @@ func TestEndpointHandlers(t *testing.T) {
 					req.Header.Set(k, v)
 				}
 			}
-			r, err := client.Do(req)
-			if err != nil {
-				t.Fatalf("failed to GET %s from component: %v", tt.path, err)
-			}
 
-			body, err := io.ReadAll(r.Body)
+			err = retry.OnError(
+				retry.DefaultBackoff,
+				func(err error) bool {
+					// the endpoint /readyz/sched-handler-sync needs some time to finish
+					// the initialization, so we need to retry
+					return !tt.useBrokenConfig &&
+						(tt.path == "/readyz" || tt.path == "/readyz/sched-handler-sync") &&
+						strings.Contains(err.Error(), "handlers are not fully synchronized")
+				},
+				func() error {
+					r, err := client.Do(req)
+					if err != nil {
+						return fmt.Errorf("failed to GET %s from component: %w", tt.path, err)
+					}
+
+					body, err := io.ReadAll(r.Body)
+					if err != nil {
+						return fmt.Errorf("failed to read response body: %w", err)
+					}
+					if err = r.Body.Close(); err != nil {
+						return fmt.Errorf("failed to close response body: %w", err)
+					}
+					if got, expected := r.StatusCode, tt.wantResponseCode; got != expected {
+						return fmt.Errorf("expected http %d at %s of component, got: %d %q", expected, tt.path, got, string(body))
+					}
+					if tt.wantResponseBodyRegx != "" {
+						matched, err := regexp.MatchString(tt.wantResponseBodyRegx, string(body))
+						if err != nil {
+							return fmt.Errorf("failed to compile regex: %w", err)
+						}
+						if !matched {
+							return fmt.Errorf("response body does not match regex.\nExpected:\n%s\n\nGot:\n%s", tt.wantResponseBodyRegx, string(body))
+						}
+					}
+					return nil
+				},
+			)
 			if err != nil {
-				t.Fatalf("failed to read response body: %v", err)
-			}
-			if err = r.Body.Close(); err != nil {
-				t.Fatalf("failed to close response body: %v", err)
-			}
-			if got, expected := r.StatusCode, tt.wantResponseCode; got != expected {
-				t.Fatalf("expected http %d at %s of component, got: %d %q", expected, tt.path, got, string(body))
-			}
-			if tt.wantResponseBodyRegx != "" {
-				matched, err := regexp.MatchString(tt.wantResponseBodyRegx, string(body))
-				if err != nil {
-					t.Fatalf("failed to compile regex: %v", err)
-				}
-				if !matched {
-					t.Fatalf("response body does not match regex.\nExpected:\n%s\n\nGot:\n%s", tt.wantResponseBodyRegx, string(body))
-				}
+				t.Fatal(err)
 			}
 		})
 	}

